### ISSUE TITLE
possible soundness problem reading Array[Short]

### ISSUE THREAD

**Issue by tpolecat** [tminglei/slick-pg#129]
Hi, I was helping someone with a problem in the `#scala` IRC channel yesterday and I think the issue was arising in slick-pg (if not it will be easy for you to verify).

In Postgres if you have a column of type `_int2` aka `smallint[]` (i.e., an array of shorts) when you read this value via `getObject` you expect to see an array containing `java.lang.Short` but instead you get an array containing `java.lang.Integer`; this may be a driver issue or a database limitation. Not sure. I discovered this in my work on [doobie](https://github.com/tpolecat/doobie).

So if you do `getObject` and extract the array values and downcast them to `java.lang.Short` you will end up with an ill-typed structure that will blow up with a `ClassCastException` when accessed at a later time. This is what the user was seeing: he was reading a Postgres `smallint[]` into a `List[Short]` and it blew up when he accessed the contents because the elements were actually `Integer`s.

Anyway it will be easy for you to verify, and I wanted you to know about this oddity in case you weren't aware. Let me know if you need more information.

**Comment by tminglei** [comment]
Hi @tpolecat, thanks for pointing it out!
Yes, I noticed it before, when you got an array value with `getObject`, sometimes you can't get you want.
But if you use `getArray`, it will be correct.

Here's the way I did in the [codes](https://github.com/tminglei/slick-pg/blob/master/core/src/main/scala/com/github/tminglei/slickpg/array/PgArrayJdbcTypes.scala#L29):
```scala
override def getValue(r: ResultSet, idx: Int): Seq[T] = {
      val value = r.getArray(idx)
      if (r.wasNull) null else value.getArray.asInstanceOf[Array[Any]].map(_.asInstanceOf[T])
    }
```

You can try it yourself.

Thanks

**Comment by tpolecat** [comment]
Sorry, I misspoke above. The problem is with `getArray`, not `getObject`. 

```scala
import java.sql.DriverManager

Class.forName("org.postgresql.Driver")

val co = DriverManager.getConnection("jdbc:postgresql:world", "postgres", "")
val ps = co.prepareStatement("select ARRAY[1,2,3]::smallint[]")
val rs = ps.executeQuery; rs.next
val ar = rs.getArray(1).getArray.asInstanceOf[Array[Any]]

ar.foreach(a => println((a, a.getClass)))

rs.close
ps.close
co.close
```

If you try this you will see that the returned array contains `Integer`, not `Short`. Again it may not be an issue for you, I just wanted to be sure you see what I'm seeing.

**Comment by tminglei** [comment]
Well, I'll try it again. And keep you updated.
_Thanks very much! :)_

**Comment by tminglei** [comment]
Hi @tpolecat, I used `int2[]` in `slick-pg`, and it looks ok.

When changed to `smallint[]`, I encountered `PSQLException: Unknown type _smallint`:
```
[error] Test com.github.tminglei.slickpg.PgArraySupportTest.testArrayFunctions failed: org.postgresql.util.PSQLException: Unknown type _smallint.
[error]     at org.postgresql.jdbc2.AbstractJdbc2Statement.setArray(AbstractJdbc2Statement.java:3063)
[error]     at com.github.tminglei.slickpg.array.PgArrayJdbcTypes$SimpleArrayJdbcType.setValue(PgArrayJdbcTypes.scala:34)
[error]     at com.github.tminglei.slickpg.array.PgArrayJdbcTypes$SimpleArrayJdbcType.setValue(PgArrayJdbcTypes.scala:21)
[error]     at com.github.tminglei.slickpg.array.PgArrayJdbcTypes$WrappedConvArrayJdbcType.setValue(PgArrayJdbcTypes.scala:119)
[error]     at com.github.tminglei.slickpg.array.PgArrayJdbcTypes$WrappedConvArrayJdbcType.setValue(PgArrayJdbcTypes.scala:110)
[error]     at scala.slick.jdbc.BaseResultConverter.set(JdbcResultConverter.scala:19)
[error]     at scala.slick.jdbc.BaseResultConverter.set(JdbcResultConverter.scala:11)
[error]     at scala.slick.relational.ProductResultConverter.set(ResultConverter.scala:71)
[error]     at scala.slick.relational.ProductResultConverter.set(ResultConverter.scala:46)
[error]     at scala.slick.relational.TypeMappingResultConverter.set(ResultConverter.scala:127)
...
```

**Comment by tminglei** [comment]
Well, I think I can reproduce it in Plain SQL array support.

If I implemented `nextShortArray`/`nextShortArrayOption` as below:
```scala
  // short array
  def nextShortArray() = nextArray[Short].getOrElse(Nil)
  def nextShortArrayOption() = nextArray[Short]
...
  ///
  private def nextArray[T](): Option[Seq[T]] = {
	val value = r.rs.getArray(r.skip.currentPos)
	if (r.rs.wasNull) None else Some(
	  value.getArray.asInstanceOf[Array[Any]].map(_.asInstanceOf[T]))
  }
```

I'll encouter an exception like this:
```
[error] Test com.github.tminglei.slickpg.PgArraySupportTest.testPlainArrayFunctions failed: java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.Short
[error]     at scala.runtime.BoxesRunTime.unboxToShort(BoxesRunTime.java:101)
[error]     at com.github.tminglei.slickpg.PgArraySupportTest$$anonfun$testPlainArrayFunctions$1$$anonfun$apply$12.apply(PgArraySupportTest.scala:236)
[error]     at com.github.tminglei.slickpg.PgArraySupportTest$$anonfun$testPlainArrayFunctions$1$$anonfun$apply$12.apply(PgArraySupportTest.scala:236)
[error]     at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
[error]     at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
[error]     at scala.collection.Iterator$class.foreach(Iterator.scala:750)
...
```

But fixing it is also easy.
When I implement it as below:
```scala
  // short array
  def nextShortArray() = nextArray[Int].getOrElse(Nil).map(_.toShort)
  def nextShortArrayOption() = nextArray[Int].map(_.map(_.toShort))
...
```

It looks ok.

**Comment by tminglei** [comment]
Anyway, it didn't cause any problem for me. _Thanks for your nice reminding! :-)_

**Comment by veegee** [comment]
I am *definitely* still getting this exact problem.

**Comment by tminglei** [comment]
Hi @veegee fixed. Pls help check, since I can't reproduce it.

**Comment by veegee** [comment]
@tminglei Yes, this fixes the problem, thanks.

**Comment by tminglei** [comment]
@veegee released slick-pg v0.12.2, pls check.

**Comment by veegee** [comment]
@tminglei Everything is working, thank you!
